konrad/cloud.py

```diff
@@ -97,27 +100,110 @@ def __init__(self, numlevels=200, num_lw_bands=16, num_sw_bands=14,
+            axis=0,
+            fill_value='extrapolate',
+        )
+        return interpolation_f
+
+    def shift_property(self, cloud_parameter, interpolation_f, z, norm_new):
+        """Shift the cloud area fraction according to a normalisation level.
+        Maintain the thickness of the cloud [m].
+
+        Parameters:
+            cloud_parameter (DataArray): cloud property to be shifted
+            interpolation_f (scipy.interpolate.interpolate.interp1d):
+                interpolation object calculated by interpolation_function
+            z (ndarray): height array [m]
+            norm_new (int / float): normalisation height [m]
+
+        Returns:
+            DataArray: shifted cloud property
+        """
+        # Move the cloud to the new normalisation level, if there is one.
+        # Otherwise keep the cloud where it is.
+        if norm_new is not np.nan:
+            cloud_parameter.values = interpolation_f(z - norm_new)
+        else:
+            cloud_parameter.values = interpolation_f(z - self._norm_level)
+
+        return cloud_parameter
+
+    def scale_optical_thickness(self, z):
+        """Conserve the total optical depth of the cloud.
+        If we assume that the cloud remains at the same altitude and with the
+        same thickness, then we also want to assume that the optical depth of
+        the cloud remains the same.
+        This is equivalent to conservation of cloud mass, as optical depth is
+        proportional to density.
+
+        Parameters:
+            z (ndarray): height array
+        """
+        dz = np.hstack([z[0], np.diff(z)])  # TODO: is this a good approx?
+        dz_sw_array = np.repeat(dz[:, np.newaxis], self.num_shortwave_bands,
+                                axis=1)
+        dz_lw_array = np.repeat(dz[:, np.newaxis], self.num_longwave_bands,
+                                axis=1)
+
+        self.longwave_optical_thickness_due_to_cloud = DataArray(
+            dz_lw_array * self._longwave_optical_thickness_per_meter,
+            dims=('mid_levels', 'num_longwave_bands'),
+            attrs={'units': 'dimensionless'}
+        )
+        self.shortwave_optical_thickness_due_to_cloud = DataArray(
+            dz_sw_array * self._shortwave_optical_thickness_per_meter,
+            dims=('mid_levels', 'num_shortwave_bands'),
+            attrs={'units': 'dimensionless'}
+        )
 
     @staticmethod
     def calculate_mass_cloud(cloud_fraction_array, z, density):
+        """Calculate the mass of the cloud on each model level given the cloud
+        area fraction and the density within the cloud.
+
+        Parameters:
+            cloud_fraction_array (ndarray): cloud area fraction
+            z (ndarray): height on each model level [m]
+            density (int / float): density of cloud mass [g m^-3]
+
+        Returns:
+            DataArray: mass of cloud on each model level [kg m^-2]
+        """
         dz = np.hstack((np.diff(z)[0], np.diff(z)))
```

Comment:
One could consider to pass the `atmosphere` object to the method and add an `Atmosphere.get_level_depth()` method. This would also ensure a consistent calculation at different locations in the code (e.g. the occurrence with the TODO comment earlier).

Reply:
Is this really a problem? Doesn't it need `z` anyhow? But I am fine with making it a function 😉